File: lib/games/poker.js
'use strict';

// ─── Texas Hold'em Poker Engine ──────────────────────────
// Server-authoritative. All deck operations happen here.
// State is serialized to boardState JSON in db.games.

const SUITS = ['s', 'h', 'd', 'c']; // spades hearts diamonds clubs
const RANKS = ['2','3','4','5','6','7','8','9','T','J','Q','K','A'];
const RANK_VAL = {};
RANKS.forEach((r, i) => { RANK_VAL[r] = i + 2; });

const HAND_RANK = {
  HIGH_CARD: 1, ONE_PAIR: 2, TWO_PAIR: 3, THREE_OF_A_KIND: 4,
  STRAIGHT: 5, FLUSH: 6, FULL_HOUSE: 7, FOUR_OF_A_KIND: 8, STRAIGHT_FLUSH: 9, ROYAL_FLUSH: 10,
};

function makeDeck() {
  const deck = [];
  for (const suit of SUITS) for (const rank of RANKS) deck.push(rank + suit);
  return deck;
}

function shuffle(deck) {
  const d = [...deck];
  for (let i = d.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [d[i], d[j]] = [d[j], d[i]];
  }
  return d;
}

function cardRank(card) { return card[0]; }
function cardSuit(card) { return card[1]; }
function cardVal(card) { return RANK_VAL[cardRank(card)]; }

function evaluateHand(cards) {
  // cards: array of 5-7 card strings, returns best 5-card evaluation
  if (cards.length < 5) return { rank: HAND_RANK.HIGH_CARD, label: 'High Card', best: cards };
  const combos = cards.length === 5 ? [cards] : choose5(cards);
  let best = null;
  for (const combo of combos) {
    const ev = evaluate5(combo);
    if (!best || ev.rank > best.rank || (ev.rank === best.rank && compareKickers(ev.kickers, best.kickers) > 0)) {
      best = ev;
    }
  }
  return best;
}

function choose5(arr) {
  const result = [];
  const n = arr.length;
  for (let a = 0; a < n - 4; a++)
    for (let b = a+1; b < n - 3; b++)
      for (let c = b+1; c < n - 2; c++)
        for (let d = c+1; d < n - 1; d++)
          for (let e = d+1; e < n; e++)
            result.push([arr[a], arr[b], arr[c], arr[d], arr[e]]);
  return result;
}

function evaluate5(cards) {
  const vals = cards.map(cardVal).sort((a, b) => b - a);
  const suits = cards.map(cardSuit);
  const isFlush = suits.every(s => s === suits[0]);
  const isStraight = checkStraight(vals);
  const counts = {};
  for (const v of vals) counts[v] = (counts[v] || 0) + 1;
  const groups = Object.entries(counts).sort((a, b) => b[1] - a[1] || b[0] - a[0]);

  let rank, label, kickers = vals;

  if (isFlush && isStraight) {
    rank = vals[0] === 14 && vals[1] === 13 ? HAND_RANK.ROYAL_FLUSH : HAND_RANK.STRAIGHT_FLUSH;
    label = rank === HAND_RANK.ROYAL_FLUSH ? 'Royal Flush' : 'Straight Flush';
    kickers = [isStraight];
  } else if (groups[0][1] === 4) {
    rank = HAND_RANK.FOUR_OF_A_KIND; label = 'Four of a Kind';
    kickers = [parseInt(groups[0][0]), parseInt(groups[1][0])];
  } else if (groups[0][1] === 3 && groups[1][1] === 2) {
    rank = HAND_RANK.FULL_HOUSE; label = 'Full House';
    kickers = [parseInt(groups[0][0]), parseInt(groups[1][0])];
  } else if (isFlush) {
    rank = HAND_RANK.FLUSH; label = 'Flush'; kickers = vals;
  } else if (isStraight) {
    rank = HAND_RANK.STRAIGHT; label = 'Straight'; kickers = [isStraight];
  } else if (groups[0][1] === 3) {
    rank = HAND_RANK.THREE_OF_A_KIND; label = 'Three of a Kind';
    kickers = [parseInt(groups[0][0]), ...vals.filter(v => v !== parseInt(groups[0][0]))];
  } else if (groups[0][1] === 2 && groups[1][1] === 2) {
    rank = HAND_RANK.TWO_PAIR; label = 'Two Pair';
    kickers = [Math.max(parseInt(groups[0][0]), parseInt(groups[1][0])),
                Math.min(parseInt(groups[0][0]), parseInt(groups[1][0])),
                ...vals.filter(v => v !== parseInt(groups[0][0]) && v !== parseInt(groups[1][0]))];
  } else if (groups[0][1] === 2) {
    rank = HAND_RANK.ONE_PAIR; label = 'One Pair';
    kickers = [parseInt(groups[0][0]), ...vals.filter(v => v !== parseInt(groups[0][0]))];
  } else {
    rank = HAND_RANK.HIGH_CARD; label = 'High Card'; kickers = vals;
  }

  return { rank, label, kickers, cards };
}

function checkStraight(sortedVals) {
  // Also handle A-2-3-4-5 (wheel)
  const uniq = [...new Set(sortedVals)];
  if (uniq.length < 5) return false;
  // Normal
  if (uniq[0] - uniq[4] === 4) return uniq[0];
  // Wheel: A-2-3-4-5
  if (uniq[0] === 14 && uniq[1] === 5 && uniq[2] === 4 && uniq[3] === 3 && uniq[4] === 2) return 5;
  return false;
}

function compareKickers(a, b) {
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return 0;
}

// ─── Game State Machine ───────────────────────────────────

const STAGE = { PREFLOP: 0, FLOP: 1, TURN: 2, RIVER: 3, SHOWDOWN: 4 };
const STAGE_NAME = ['preflop', 'flop', 'turn', 'river', 'showdown'];

function createPokerState(players, stakeKas, options = {}) {
  // players: [{ addr, name }]
  const ante = options.ante || 0;
  const smallBlind = options.smallBlind || Math.max(1, Math.floor(stakeKas * 0.05));
  const bigBlind = smallBlind * 2;
  const startingChips = options.startingChips || Math.floor(stakeKas * 10); // 10x buy-in in chips
  const deck = shuffle(makeDeck());

  const playerStates = players.map((p, i) => ({
    addr: p.addr,
    name: p.name || 'Player ' + (i + 1),
    chips: startingChips,
    holeCards: [deck.pop(), deck.pop()],
    bet: 0,
    totalBet: 0,
    folded: false,
    allIn: false,
    isDealer: i === 0,
    isSB: i === (players.length === 2 ? 0 : 1),
    isBB: i === (players.length === 2 ? 1 : 2 % players.length),
    handRank: null,
  }));

  // Post blinds
  const sbIdx = players.length === 2 ? 0 : 1;
  const bbIdx = players.length === 2 ? 1 : 2 % players.length;
  playerStates[sbIdx].chips -= smallBlind;
  playerStates[sbIdx].bet = smallBlind;
  playerStates[sbIdx].totalBet = smallBlind;
  playerStates[bbIdx].chips -= bigBlind;
  playerStates[bbIdx].bet = bigBlind;
  playerStates[bbIdx].totalBet = bigBlind;

  // UTG acts first preflop
  const firstToAct = (bbIdx + 1) % players.length;

  return {
    stage: STAGE.PREFLOP,
    stageName: 'preflop',
    deck,  // remaining deck (server-only; strip before sending to clients)
    community: [],
    pot: smallBlind + bigBlind,
    currentBet: bigBlind,
    minRaise: bigBlind,
    players: playerStates,
    activePlayerIdx: firstToAct,
    dealerIdx: 0,
    sbIdx,
    bbIdx,
    smallBlind,
    bigBlind,
    lastRaiserIdx: bbIdx,
    actedThisRound: new Array(players.length).fill(false),
    handNumber: 1,
    winner: null,
    winReason: null,
    finished: false,
  };
}

function getPublicState(state, viewerAddr) {
  // Strip hole cards from opponents
  return {
    ...state,
    deck: undefined,
    players: state.players.map(p => ({
      ...p,
      holeCards: (p.addr === viewerAddr || state.stage === STAGE.SHOWDOWN) ? p.holeCards : p.holeCards.map(() => 'back'),
    })),
  };
}

function applyAction(state, playerAddr, action, amount) {
  const idx = state.players.findIndex(p => p.addr === playerAddr);
  if (idx === -1) return { error: 'Player not found' };
  if (idx !== state.activePlayerIdx) return { error: 'Not your turn' };
  const player = state.players[idx];
  if (player.folded) return { error: 'Already folded' };

  const s = JSON.parse(JSON.stringify(state)); // deep clone
  const p = s.players[idx];

  switch (action) {
    case 'fold':
      p.folded = true;
      s.actedThisRound[idx] = true;
      break;

    case 'check':
      if (s.currentBet > p.bet) return { error: 'Cannot check, must call or raise' };
      s.actedThisRound[idx] = true;
      p.bet = 0;
      break;

    case 'call': {
      const callAmount = Math.min(s.currentBet - p.bet, p.chips);
      p.chips -= callAmount;
      p.totalBet += callAmount;
      s.pot += callAmount;
      if (p.chips === 0) p.allIn = true;
      s.actedThisRound[idx] = true;
      p.bet = s.currentBet;
      break;
    }

    case 'raise': {
      if (!amount || amount < s.minRaise) return { error: 'Raise must be at least ' + s.minRaise };
      const toCall = s.currentBet - p.bet;
      const total = toCall + amount;
      if (total > p.chips) return { error: 'Not enough chips' };
      p.chips -= total;
      p.totalBet += total;
      s.pot += total;
      s.currentBet += amount;
      s.minRaise = amount;
      s.lastRaiserIdx = idx;
      s.actedThisRound = new Array(s.players.length).fill(false);
      s.actedThisRound[idx] = true;
      p.bet = s.currentBet;
      if (p.chips === 0) p.allIn = true;
      break;
    }

    case 'allin': {
      const allInAmount = p.chips;
      p.totalBet += allInAmount;
      s.pot += allInAmount;
      if (allInAmount + p.bet > s.currentBet) {
        s.currentBet = allInAmount + p.bet;
        s.lastRaiserIdx = idx;
        s.actedThisRound = new Array(s.players.length).fill(false);
      }
      p.bet = s.currentBet;
      p.chips = 0;
      p.allIn = true;
      s.actedThisRound[idx] = true;
      break;
    }

    default:
      return { error: 'Unknown action: ' + action };
  }

  // Check if round is over
  const activePlayers = s.players.filter(p => !p.folded && !p.allIn);
  const foldedCount = s.players.filter(p => p.folded).length;

  // Everyone folded except one
  if (foldedCount === s.players.length - 1) {
    const winner = s.players.find(p => !p.folded);
    winner.chips += s.pot;
    s.winner = winner.addr;
    s.winReason = 'fold';
    s.finished = true;
    s.deck = undefined;
    return { state: s, finished: true, winner: winner.addr, reason: 'fold' };
  }

  // Advance to next active player
  let next = (idx + 1) % s.players.length;
  while (s.players[next].folded || s.players[next].allIn) {
    next = (next + 1) % s.players.length;
    if (next === idx) break;
  }

  // Check if betting round is complete
  const roundDone = isBettingRoundDone(s);
  if (roundDone) {
    return advanceStage(s);
  }

  s.activePlayerIdx = next;
  s.deck = undefined; // never serialize deck to client
  return { state: s };
}

function isBettingRoundDone(state) {
  const active = state.players.filter(p => !p.folded && !p.allIn);
  if (active.length === 0) return true;
  // All active players have acted and bets are equal
  return active.every((p, i) => {
    const idx = state.players.indexOf(p);
    return state.actedThisRound[idx] && p.bet === state.currentBet;
  });
}

function advanceStage(state) {
  const s = state;
  // Reset bets
  s.players.forEach(p => { p.bet = 0; });
  s.currentBet = 0;
  s.minRaise = s.bigBlind;
  s.actedThisRound = new Array(s.players.length).fill(false);
  s.lastRaiserIdx = -1;

  if (s.stage === STAGE.PREFLOP) {
    // Deal flop: need deck — but deck was stripped. Re-attach from stored state if needed.
    // In practice the server reconstructs from the stored deck.
    s.stage = STAGE.FLOP;
    s.stageName = 'flop';
    s.community.push(s._deck?.pop() || 'Xs', s._deck?.pop() || 'Xs', s._deck?.pop() || 'Xs');
  } else if (s.stage === STAGE.FLOP) {
    s.stage = STAGE.TURN;
    s.stageName = 'turn';
    s.community.push(s._deck?.pop() || 'Xs');
  } else if (s.stage === STAGE.TURN) {
    s.stage = STAGE.RIVER;
    s.stageName = 'river';
    s.community.push(s._deck?.pop() || 'Xs');
  } else if (s.stage === STAGE.RIVER) {
    return goToShowdown(s);
  }

  // First to act after dealer (skip folded/allIn)
  let next = (s.dealerIdx + 1) % s.players.length;
  while (s.players[next].folded || s.players[next].allIn) {
    next = (next + 1) % s.players.length;
  }
  s.activePlayerIdx = next;
  s.deck = undefined;
  return { state: s };
}

function goToShowdown(state) {
  const s = state;
  s.stage = STAGE.SHOWDOWN;
  s.stageName = 'showdown';

  const activePlayers = s.players.filter(p => !p.folded);
  for (const p of activePlayers) {
    const allCards = [...p.holeCards, ...s.community];
    p.handRank = evaluateHand(allCards);
  }

  activePlayers.sort((a, b) => {
    if (a.handRank.rank !== b.handRank.rank) return b.handRank.rank - a.handRank.rank;
    return compareKickers(b.handRank.kickers, a.handRank.kickers);
  });

  const winner = activePlayers[0];
  const isTie = activePlayers.length > 1 &&
    activePlayers[1].handRank.rank === winner.handRank.rank &&
    compareKickers(activePlayers[1].handRank.kickers, winner.handRank.kickers) === 0;

  if (isTie) {
    const splitPot = Math.floor(s.pot / activePlayers.length);
    activePlayers.forEach(p => { const orig = s.players.find(x => x.addr === p.addr); orig.chips += splitPot; });
    s.winner = 'draw';
    s.winReason = 'split pot';
  } else {
    const winnerOrig = s.players.find(p => p.addr === winner.addr);
    winnerOrig.chips += s.pot;
    s.winner = winner.addr;
    s.winReason = winner.handRank.label;
  }

  s.finished = true;
  s.deck = undefined;
  return { state: s, finished: true, winner: s.winner, reason: s.winReason };
}

// ─── Server-side state manager (keeps deck private) ───────
class PokerEngine {
  constructor() {
    this.games = new Map(); // gameId -> { state, deck }
  }

  startGame(gameId, players, stakeKas, options) {
    const state = createPokerState(players, stakeKas, options);
    const deck = state.deck;
    state._deck = deck; // keep deck reference
    this.games.set(gameId, { state, deck });
    return this.getPublicState(gameId);
  }

  applyAction(gameId, playerAddr, action, amount) {
    const entry = this.games.get(gameId);
    if (!entry) return { error: 'Game not found' };
    const result = applyAction(entry.state, playerAddr, action, amount);
    if (result.error) return result;
    // Restore deck on state for continued play
    result.state._deck = entry.deck;
    entry.state = result.state;
    return result;
  }

  getPublicState(gameId, viewerAddr) {
    const entry = this.games.get(gameId);
    if (!entry) return null;
    return getPublicState(entry.state, viewerAddr);
  }

  dealNextStreet(gameId) {
    const entry = this.games.get(gameId);
    if (!entry) return { error: 'Game not found' };
    const s = entry.state;
    // Pop from the stored deck
    if (s.stage === STAGE.PREFLOP) {
      entry.deck.pop(); // burn
      s.community.push(entry.deck.pop(), entry.deck.pop(), entry.deck.pop());
      s.stage = STAGE.FLOP; s.stageName = 'flop';
    } else if (s.stage === STAGE.FLOP) {
      entry.deck.pop(); // burn
      s.community.push(entry.deck.pop());
      s.stage = STAGE.TURN; s.stageName = 'turn';
    } else if (s.stage === STAGE.RIVER) {
      entry.deck.pop(); // burn
      s.community.push(entry.deck.pop());
      s.stage = STAGE.RIVER; s.stageName = 'river';
    }
    return this.getPublicState(gameId);
  }

  endGame(gameId) {
    this.games.delete(gameId);
  }
}

module.exports = {
  PokerEngine,
  createPokerState,
  applyAction,
  evaluateHand,
  getPublicState,
  STAGE,
  HAND_RANK,
  RANKS,
  SUITS,
};

File: crates/htp-games/src/poker.rs
//! High Table Protocol - Texas Hold'em Poker Engine
//! Ported from 456-line JS engine. 2-player, blinds, full hand evaluation, betting rounds.

use serde::{Deserialize, Serialize};
use rand::seq::SliceRandom;
use crate::{GameError, GameOutcome, GameStatus};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct PokerCard { pub rank: u8, pub suit: u8 }

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum HandRank { HighCard=0, OnePair, TwoPair, ThreeOfKind, Straight, Flush, FullHouse, FourOfKind, StraightFlush, RoyalFlush }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HandEval { pub rank: HandRank, pub kickers: Vec<u8>, pub cards: Vec<PokerCard> }

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum PokerStage { Preflop=0, Flop, Turn, River, Showdown }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PokerAction { Fold, Check, Call, Raise(u32), AllIn }

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PokerActionResult {
    pub state: Option<PokerState>,
    pub finished: bool,
    pub winner: Option<String>,
    pub reason: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PokerPlayer { pub addr: String, pub name: String, pub chips: u32, pub hole_cards: Vec<PokerCard>, pub bet: u32, pub total_bet: u32, pub folded: bool, pub all_in: bool, pub hand_rank: Option<HandEval> }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PokerState {
    pub stage: PokerStage, pub stage_name: String, pub community: Vec<PokerCard>,
    pub pot: u32, pub current_bet: u32, pub min_raise: u32,
    pub players: Vec<PokerPlayer>, pub active_player_idx: usize,
    pub dealer_idx: usize, pub small_blind: u32, pub big_blind: u32,
    pub last_raiser_idx: Option<usize>, pub acted_this_round: Vec<bool>,
    pub hand_number: u32, pub finished: bool,
    #[serde(skip)] pub deck: Vec<PokerCard>,
    #[serde(skip)] sb_idx: usize, #[serde(skip)] bb_idx: usize,
    finished_reason: Option<String>,
}

fn make_deck() -> Vec<PokerCard> {
    (2u8..=14).flat_map(|rank| (0..4).map(move |suit| PokerCard{rank,suit})).collect()
}

fn card_val(c: &PokerCard) -> u8 { c.rank }

fn choose_combos(cards: &[PokerCard]) -> Vec<Vec<PokerCard>> {
    let n = cards.len(); if n==5 {return vec![cards.to_vec()];}
    let mut out=vec![];
    for a in 0..n-4{for b in a+1..n-3{for c in b+1..n-2{for d in c+1..n-1{for e in d+1..n{
        out.push(vec![cards[a],cards[b],cards[c],cards[d],cards[e]]);
    }}}}}
    out
}

fn check_straight(vals: &[u8]) -> Option<u8> {
    let mut u: Vec<u8>=vals.to_vec(); u.dedup(); if u.len()<5{return None;}
    if u[0]-u[4]==4{return Some(u[0]);}
    if u[0]==14&&u[1]==5&&u[2]==4&&u[3]==3&&u[4]==2{return Some(5);}
    None
}

fn evaluate5(cards: &[PokerCard]) -> HandEval {
    let mut vals: Vec<u8>=cards.iter().map(card_val).collect();
    vals.sort_unstable_by(|a,b|b.cmp(a));
    let flush=cards.iter().all(|c|c.suit==cards[0].suit);
    let sh=check_straight(&vals); let straight=sh.is_some();
    let mut cnt=HashMap::new();
    for v in &vals{*cnt.entry(*v).or_default()+=1;}
    let mut g: Vec<(u8,u8)>=cnt.iter().map(|(&k,&v)|(k,v)).collect();
    g.sort_by(|a,b|b.1.cmp(&a.1).then(b.0.cmp(&a.0)));

    let(rank,kickers)=match(flush,straight,g.as_slice()){
        (true,true,_) if vals[0]==14&&vals[1]==13=>(HandRank::RoyalFlush,vec![sh.unwrap()]),
        (true,true,_)=>(HandRank::StraightFlush,vec![sh.unwrap()]),
        (_,_,[(_,4),(o,1),..])=>(HandRank::FourOfKind,vec![g[0].0,*o]),
        (_,_,[(t,3),(p,2),..])=>(HandRank::FullHouse,vec![*t,*p]),
        (true,false,_)=>(HandRank::Flush,vals.clone()),
        (false,true,_)=>(HandRank::Straight,vec![sh.unwrap()]),
        (_,_,[(t,3),..])=>{let mut k=vec![*t];k.extend(vals.iter().filter(|&&v|v!=*t));k;(HandRank::ThreeOfKind,k)}.1,
        (_,_,[(p1,2),(p2,2),(k,1),..])=>(HandRank::TwoPair,vec![*p1.max(p2),*p1.min(p2),*k]),
        (_,_,[(p,2),..])=>{let mut k=vec![*p];k.extend(vals.iter().filter(|&&v|v!=*p));k;(HandRank::OnePair,k)}.1,
        _=>(HandRank::HighCard,vals.clone()),
    };
    HandEval{rank,kickers,cards:cards.to_vec()}
}

fn cmp_kickers(a:&[u8],b:&[u8])->std::cmp::Ordering{
    for(x,y)in a.iter().zip(b.iter()){let o=x.cmp(y);if o!=std::cmp::Ordering::Equal{return o;}}
    a.len().cmp(&b.len())
}

pub fn evaluate_hand(cards:&[PokerCard])->HandEval{
    if cards.len()<5{return HandEval{rank:HandRank::HighCard,kickers:cards.iter().map(card_val).rev().collect(),cards:cards.to_vec()};}
    let combos=choose_combos(cards); let mut best:Option<HandEval>=None;
    for c in combos{let ev=evaluate5(&c); if let Some(ref b)=best{if ev.rank>b.rank||(ev.rank==b.rank&&cmp_kickers(&ev.kickers,&b.kickers)==std::cmp::Ordering::Greater){best=Some(ev);}}else{best=Some(ev);}}
    best.unwrap()
}

impl PokerState{
    pub fn new(players:Vec<(String,String)>,stake_kas:u32,sb:u32,bb:u32)->Self{
        let chips=stake_kas*10; let mut d=make_deck();
        {let mut rng=rand::thread_rng(); d.shuffle(&mut rng);}
        let sb_i=0; let bb_i=1;
        let mut ps:Vec<PokerPlayer>=players.into_iter().enumerate().map(|(i,(a,n))|{
            PokerPlayer{addr:a,name:n,chips,hole_cards:vec![d.pop().unwrap(),d.pop().unwrap()],bet:0,total_bet:0,folded:false,all_in:false,hand_rank:None}
        }).collect();
        ps[sb_i].chips=ps[sb_i].chips.saturating_sub(sb); ps[sb_i].bet=sb; ps[sb_i].total_bet=sb;
        ps[bb_i].chips=ps[bb_i].chips.saturating_sub(bb); ps[bb_i].bet=bb; ps[bb_i].total_bet=bb;
        let n=ps.len();
        PokerState{stage:PokerStage::Preflop,stage_name:"preflop".into(),deck:d,community:vec![],pot:sb+bb,current_bet:bb,min_raise:bb,players:ps,active_player_idx:(bb_i+1)%n,dealer_idx:0,sb_idx:sb_i,bb_idx:bb_i,small_blind:sb,big_blind:bb,last_raiser_idx:Some(bb_i),acted_this_round:vec![false;n],hand_number:1,finished:false,finished_reason:None}
    }

    pub fn to_public(&self,viewer:&str)->Self{
        let mut s=self.clone();
        for p in &mut s.players{if p.addr!=viewer&&self.stage!=PokerStage::Showdown{p.hole_cards=vec![PokerCard{rank:0,suit:0};2];}}
        s.deck=vec![]; s
    }

    pub fn apply_action(&mut self,action:&PokerAction)->PokerActionResult{
        if self.finished{return PokerActionResult{error:Some("hand over".into()),finished:true,..Default::default()};}
        let i=self.active_player_idx;
        let p=&mut self.players[i];
        let amt=match action{PokerAction::Raise(a)=>*a,PokerAction::AllIn=>p.chips,_=>0};
        match action{
            PokerAction::Fold=>{p.folded=true;self.acted_this_round[i]=true;}
            PokerAction::Check=>{if self.current_bet!=p.bet{return PokerActionResult{error:Some("must call or raise".into()),..Default::default()};}self.acted_this_round[i]=true;}
            PokerAction::Call=>{let call=self.current_bet-p.bet;let a=call.min(p.chips);p.chips-=a;p.bet+=a;p.total_bet+=a;self.pot+=a;if p.chips==0{p.all_in=true;}self.acted_this_round[i]=true;}
            PokerAction::Raise(_)=>{
                if amt<self.min_raise&&p.chips>=self.min_raise+self.current_bet-p.bet{return PokerActionResult{error:Some(format!("min raise {}",self.min_raise)),..Default::default()};}
                let need=amt+self.current_bet-p.bet;let a=need.min(p.chips);p.chips-=a;p.bet+=a;p.total_bet+=a;self.pot+=a;self.current_bet=p.bet;self.min_raise=amt;self.last_raiser_idx=Some(i);if p.chips==0{p.all_in=true;}self.acted_this_round=vec![false;self.players.len()];self.acted_this_round[i]=true;
            }
            PokerAction::AllIn=>{let a=p.chips;p.chips=0;p.bet+=a;p.total_bet+=a;self.pot+=a;p.all_in=true;if p.bet>self.current_bet{self.current_bet=p.bet;let ra=p.bet-self.current_bet+a;if ra>=self.min_raise{self.min_raise=ra;self.last_raiser_idx=Some(i);self.acted_this_round=vec![false;self.players.len()];}}self.acted_this_round[i]=true;}
        }
        let active:Vec<_>=self.players.iter().filter(|p|!p.folded).collect();
        if active.len()==1{let wa=active[0].addr.clone();self.players.iter_mut().find(|p|p.addr==wa).unwrap().chips+=self.pot;self.finished=true;self.finished_reason=Some("fold".into());return PokerActionResult{finished:true,winner:Some(wa),reason:Some("fold".into()),..Default::default()};}
        if self.is_round_done(){return self.advance_stage();}
        let mut nxt=(i+1)%self.players.len();while self.players[nxt].folded||self.players[nxt].all_in{nxt=(nxt+1)%self.players.len();if nxt==i{break;}}self.active_player_idx=nxt;
        PokerActionResult{state:Some(self.clone()),..Default::default()}
    }

    fn is_round_done(&self)->bool{
        let a:Vec<_>=self.players.iter().enumerate().filter(|(_,p)|!p.folded&&!p.all_in).collect();
        if a.is_empty(){return true;} a.iter().all(|(i,p)|self.acted_this_round[*i]&&p.bet==self.current_bet)
    }

    fn advance_stage(&mut self)->PokerActionResult{
        for p in &mut self.players{p.bet=0;} self.current_bet=0; self.min_raise=self.big_blind; self.acted_this_round=vec![false;self.players.len()]; self.last_raiser_idx=None;
        match self.stage{
            PokerStage::Preflop=>{self.deal(3);self.stage=PokerStage::Flop;self.stage_name="flop".into();}
            PokerStage::Flop=>{self.deal(1);self.stage=PokerStage::Turn;self.stage_name="turn".into();}
            PokerStage::Turn=>{self.deal(1);self.stage=PokerStage::River;self.stage_name="river".into();}
            PokerStage::River=>return self.showdown(),
            PokerStage::Showdown=>return PokerActionResult{error:Some("already showdown".into()),finished:true,..Default::default()},
        }
        let mut n=(self.dealer_idx+1)%self.players.len();while self.players[n].folded||self.players[n].all_in{n=(n+1)%self.players.len();}self.active_player_idx=n;
        PokerActionResult{state:Some(self.clone()),..Default::default()}
    }

    fn deal(&mut self,n:usize){self.deck.pop();for _ in 0..n{if let Some(c)=self.deck.pop(){self.community.push(c);}}}

    fn showdown(&mut self)->PokerActionResult{
        self.stage=PokerStage::Showdown;self.stage_name="showdown".into();
        for p in &mut self.players{if p.folded{continue;}let ac:Vec<PokerCard>=p.hole_cards.iter().chain(self.community.iter()).copied().collect();p.hand_rank=Some(evaluate_hand(&ac));}
        let mut a:Vec<usize>=self.players.iter().enumerate().filter(|(_,p)|!p.folded).map(|(i,_)|i).collect();
        a.sort_by(|&x,&y|{
            let rx=self.players[x].hand_rank.as_ref().map(|e|e.rank).unwrap_or(HandRank::HighCard);
            let ry=self.players[y].hand_rank.as_ref().map(|e|e.rank).unwrap_or(HandRank::HighCard);
            match ry.cmp(&rx){std::cmp::Ordering::Equal=>{let kx=self.players[x].hand_rank.as_ref().map(|e|&*e.kickers);let ky=self.players[y].hand_rank.as_ref().map(|e|&*e.kickers);match(kx,ky){(Some(a),Some(b))=>cmp_kickers(b,a),_=>std::cmp::Ordering::Equal,}}o=>o,}
        });
        let w=&self.players[a[0]];let wa=w.addr.clone();let r=w.hand_rank.as_ref().map(|e|format!("{:?}",e.rank)).unwrap_or_default();
        self.players[a[0]].chips+=self.pot;self.finished=true;self.finished_reason=Some(r.clone());
        PokerActionResult{state:Some(self.clone()),finished:true,winner:Some(wa),reason:Some(r),..Default::default()}
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PokerGame { pub state: PokerState, pub status: GameStatus, pub outcome: GameOutcome }

impl PokerGame{
    pub fn new(a:String,b:String,stake:u32,sb:u32,bb:u32)->Self{
        PokerGame{state:PokerState::new(vec![(a,"P1".into()),(b,"P2".into())],stake,sb,bb),status:GameStatus::Active,outcome:GameOutcome::Pending}
    }
    pub fn action(&mut self,addr:&str,act:&PokerAction)->PokerActionResult{
        let cur=&self.state.players[self.state.active_player_idx].addr;
        if cur!=addr{return PokerActionResult{error:Some(format!("not your turn — {}",cur)),..Default::default()};}
        let r=self.state.apply_action(act);
        if r.finished{self.status=GameStatus::Complete;self.outcome=match r.winner.as_deref(){
            Some(w) if *w==self.state.players[0].addr=>GameOutcome::Player1Wins,
            Some(_)=>GameOutcome::Player2Wins,None=>GameOutcome::Draw,};}
        r
    }
}

impl Default for PokerGame{fn default()->Self{Self::new("".into(),"".into(),10,1,2)}}
